libdevice/fallback-cstring.cpp

```diff
@@ -7,9 +7,89 @@
 //===----------------------------------------------------------------------===//
 
 #include "wrapper.h"
+#include <cstdint>
 
 #ifdef __SPIR__
+DEVICE_EXTERN_C
 void *__devicelib_memcpy(void *dest, const void *src, size_t n) {
   return __builtin_memcpy(dest, src, n);
 }
+
+DEVICE_EXTERN_C
+void *__devicelib_memset(void *dest, int c, size_t n) {
+  return __builtin_memset(dest, c, n);
+}
+
+static int __devicelib_memcmp_1byte(const void *s1, const void *s2, size_t n) {
+  const uint8_t *s1_uint8_ptr = reinterpret_cast<const uint8_t *>(s1);
+  const uint8_t *s2_uint8_ptr = reinterpret_cast<const uint8_t *>(s2);
+  while (n > 0) {
+    if (*s1_uint8_ptr == *s2_uint8_ptr) {
+      s1_uint8_ptr++;
+      s2_uint8_ptr++;
+      n--;
+    } else {
+      return *s1_uint8_ptr - *s2_uint8_ptr;
+    }
+  }
+
+  return 0;
+}
+
+static int __devicelib_memcmp_4byte(const void *s1, const void *s2, size_t n) {
+  const uint32_t *s1_uint32_ptr = reinterpret_cast<const uint32_t *>(s1);
+  const uint32_t *s2_uint32_ptr = reinterpret_cast<const uint32_t *>(s2);
+  while (n >= 4) {
+    if (*s1_uint32_ptr == *s2_uint32_ptr) {
+      s1_uint32_ptr++;
+      s2_uint32_ptr++;
+      n -= 4;
+    } else {
+      n = 4;
+      break;
+    }
+  }
+
+  const uint8_t *s1_uint8_ptr =
```

Comment:
May __devicelib_memcmp_1byte be reused here?

Reply:
Hi, @mlychkov 
Yes, already updated the patch.
Thanks very much.